sdk/core/azure-core/inc/credentials/credentials.hpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <chrono>
+#include <mutex>
+#include <string>
+
+namespace azure
+{
+namespace core
+{
+namespace credentials
+{
+
+class Credential
+{
+  virtual void SetScopes(std::string const& scopes) { (void)scopes; }
+
+public:
+  class Internal;
+  virtual ~Credential() noexcept = default;
+
+protected:
+  Credential() = default;
+
+  Credential(Credential const& other) = default;
+  Credential& operator=(Credential const& other) = default;
+};
+
+class TokenCredential : public Credential
+{
+  class Token
+  {
+    friend class TokenCredential;
+
+  public:
+    std::string Scopes;
+    std::string TokenString;
+    std::chrono::system_clock::time_point ExpiresAt;
+
+  private:
+    Token() {}
+    explicit Token(std::string const& scopes) : Scopes(scopes) {}
+
+    Token(
```

Comment:
Have you considered just letting this type be an aggregate?

```
struct Token {
std::string Scopes;
std::string TokenString;
std::chrono::system_clock::time_point ExpiresAt;
};
```

Reply:
Done, thanks! At first, I did not like the idea because there are no designators until C++ 20. But then, you can mismatch the order of string  arguments in constructor as well.
I may revert it back to being a class in the future, if there is more complex logic needed, but in its current incarnation, aggregate is fine.